c++/src/vector.cc

```diff
@@ -0,0 +1,100 @@
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <initializer_list>
+#include <cassert>
+
+template <class T>
+class Vector {
+private:
+  unsigned int length = 0;
+  unsigned int capacity = 8;
+  T* dataStore = nullptr;
+
+public:
+  Vector() : dataStore(new T[this->capacity]{0}) {}
+
+  Vector(std::initializer_list<T> list) 
+  : dataStore(new T[this->capacity]{0}) {
+    for (const T &item : list) {
+      this->push(item);
+    }
+  }
+
+  ~Vector() {
+    delete[] this->dataStore;
+  }
+
+
+  void push(T val) {
+    if (length >= capacity) {
+      std::cout << "Vector capacity exceeded. ";
+      std::exit(1);
+    }
+
+    this->dataStore[this->length] = val;
+    this->length++;
+  }
+
+  T pop() {
+    if (length == 0) {
+      std::cout << "Attempted to pop empty vector. ";
+      std::exit(1);
+    }
+
+    length--;
+    return this->dataStore[length + 1];
+  }
+
+  void toOstream(std::ostream &stream) {
+    for (unsigned int i = 0; i < this->length; i++) {
+      stream << this->dataStore[i];
+
+      if (i < this->length - 1) {
+        stream << " → ";
+      }
+    }
+  }
+};
+
+template <typename T>
+void test_expected_values(Vector<T> &vector, std::string expectedString);
+
+void test_push() {
```

Comment:
Good work here. However, it's preferable to try to keep one "test" per test. So right here you have two. The idea you want to aim for is that if a test crashes, it's for one reason and one reason only, so you right away know what's wrong. Here there are two tests. If they're testing the same thing, i.e. they would crash for the same reason, then it's just wasted time on your part to write two: they're redundant, and they don't _technically_ make things safer, they just make you feel more at ease, but _technically_/_practically_ it's not necessary, kind of like earlier how "just in case I'll use `>=`", except worse here cause it entails unnecessary work.

So instead of a very broad `test_push` I would create `test_push_when_full()`, `test_push_when_empty()`, etc. if you want. You can still keep `test_push` as a way of grouping the push tests if you want, so in `test_push` you would call `test_push_when_full` et al, then in main you just call `test_push`. The benefit of this is that if you only want to focus on pop tests for whatever reason, you would be able to just comment out `test_push()` and leave `test_pop()`.